Expose Flush API

ryansmith3136 opened the pull request:
This PR accomplishes two things:
1. Expose a flush API such that we have API similarity to Ruby's StringIO class.
2. Introduce fine grained locks which will make the system more reasonable.

These changes were primarily motivated by this issue: ryandotsmith/lpxc#7

ryansmith3136 committed:
refactor: msgs is in fact a queue.

ryansmith3136 committed:
Make hashed queues thread safe.

Each queue now will expose an API to determine if it is full
and also an API to drain all messages from the queue.  This change does
not impact current behaviour and is a step forward towards creating fine
grained locks for lpxc such that we can expose a flush API for the client.

ryansmith3136 commented:
Related: heroku/heroku-buildpack-ruby#192 and ryandotsmith/lpxc#7

ryansmith3136 committed:
Feature: expose `flush` in API

Lpxc#flush is now thread safe. In order to make it thread safe, we had to
synchronize a mutex inside the method. This meant that we had to remove
the call to flush in Lpxc#puts outside of the mutex in which we initalized the
FlushableLockableSizedQueue. Thus avoiding deadlock.

In addition, this commit removed the external locking in Lpxc#delay_flush
since the flush method is now responsible for such activity.

ryansmith3136 committed:
make a note about how the client will auto-flush

ryansmith3136 committed:
add a test for the Lpxc#flush

eric commented on lib/lpxc.rb:
```diff
@@ -4,6 +4,31 @@
 
+class LockableSizedQueue < SizedQueue
+  def initialize(max)
+    @locker = Mutex.new
```
Why introduce a second mutex instead of the one that is already in `SizedQueue`?

ryansmith3136 replied:
Great question. My original approach involved trying to re-use the mutex inside of `SizedQueue`. However, I am pretty sure that to use it I would have to preform a trick to gain access to the private variable. I found that a tad ironic given the problems outlined in #7. Furthermore, it wasn't clear to me that it was safe for me to use the mutex inside `SizedQueue`. Thus I ended up adding this mutex.

Can you outline an approach that would be both safe and reasonable using the mutex inside `SizedQueue`?

eric replied:
You may just consider implementing this queue yourself with a mutex and an array. The implementation of `SizedQueue` is simple enough that you may not really buy anything by using it. The blocking aspect of `SizedQueue` isn't even really buying you anything here.

eric commented on lib/lpxc.rb:
```diff
@@ -76,38 +103,37 @@ def wait
+      @hash.each do |tok, queue|
+        #Copy the messages from the queue into the payload array.
+        payloads = queue.flush
+        return if payloads.nil? || payloads.empty?
```
Is this `return` supposed to be a `next`?

ryansmith3136 replied:
I think you are correct. Allow me to fix this problem. This bug has been around for a while now. I suspect that it is rarely the case payloads is nil or empty.

ryansmith3136 committed:
Refactor: move full method into flushable queue

ryansmith3136 committed:
refactor: self is superfluous

ryansmith3136 committed:
bugfix: continue processing hash when one queue is empty

This bug has been around for a while now. I suspect that it is rarely the case payloads is nil or empty.

eric commented on lib/lpxc.rb:
```diff
@@ -76,38 +103,37 @@ def wait
-        body += "#{fmt(payload)}"
-      end
+    @hash_lock.synchronize do
+      @hash.each do |tok, queue|
```
Why not just replace `@hash` with a new instance that is empty here?

It would allow you to release the lock and process the old one outside of the lock.

eric commented on lib/lpxc.rb:
```diff
@@ -54,16 +79,18 @@ def initialize(opts={})
-      #the queue if we add the last item.
-      q.enq({:t => Time.now.utc, :token => tok, :msg => msg})
-      flush if q.size == q.max
+      q = @hash[tok] ||= FlushableLockableSizedQueue.new(@batch_size)
```
I generally like to move stuff like this to a separate method (with the synchronization) to isolate initialization and synchronization from everything else.

ryansmith3136 committed:
Refactor: use specific data structure

This is more clear as we build a data structure specifically for
our use case.

ryansmith3136 committed:
Refactor: process hash outside of lock

We are replacing the instance of @hash inside the synchronized block
which which decrease the amount of time we need to operate in a shared lock.

ryansmith3136 commented:
@eric Want to take a look at the `@hash` processing routine? I think I implemented the solution you described.

ryansmith3136 committed:
move processing into method

eric commented on lib/lpxc.rb:
```diff
@@ -4,6 +4,26 @@
+  end
+
+  def flush
+    @locker.synchronize {@array.size.times.map {@array.pop}}
```
Why not just return the current `@array` and allocate a new one? I would guess that would create less new allocations.

eric commented on lib/lpxc.rb:
```diff
@@ -4,6 +4,26 @@
+  end
+
+  def full?
+    @locker.synchronize {@array.size == @max}
```
Might be safer to make that `>=`.

eric replied:
Actually, looking at the code, you definitely need to make it `>=` because there is no synchronization between the `enqueue` and `full?`.

ryansmith3136 committed:
Add a more liberal condition for determining LogMsgQueue fullness

ryansmith3136 committed:
Simpler flush implementation for LogMsgQueue

schneems commented on lib/lpxc.rb:
```diff
@@ -54,16 +78,18 @@ def initialize(opts={})
+    #This call will block if the queue is full.
+    #However this should never happen since the next command will flush
+    #the queue if we add the last item.
+    q.enqueue({:t => Time.now.utc, :token => tok, :msg => msg})
```
> This call will block if the queue is full.

is that true? Looks like it just appends to the `@array`. We cannot count on the relationship between `q.enqueue` and `q.full?` as the two are not tied together or guaranteed to be atomic, you could have `q.enqueue` on the same queue called via a hundred threads before `flush if q.full?` is called.

ryansmith3136 replied:
You are correct. That comment is inaccurate. It was accurate when we were using a SizedQueue. I think the comment should be removed as I don't see an advantage to blocking.

schneems commented on lib/lpxc.rb:
```diff
@@ -54,16 +78,18 @@ def initialize(opts={})
+    #the queue if we add the last item.
+    q.enqueue({:t => Time.now.utc, :token => tok, :msg => msg})
+    # Consequently we flush the entire client if any queue is full.
+    flush if q.full?
```
`flush if q.full?` is not atomic, is that okay in this case, it may be possible that `if q.full?` gets called and then another thread interrupts

ryansmith3136 replied:
I am wondering if this is an acceptable situation. I am leaning towards yes in favor of us optimizing for less locking/blocking. The downside is that we may end up with more messages in our queue than originally agreed upon via the batch size parameter. I think that is ok as long as it is documented.

schneems replied:
We're already hitting a synchronize for `q.full?` what if we have it yield to a block when present 

```
q.full? do
  flush
end
```

Maybe alias `full?` to `when_full` or something else to make the code more human readable? `q.when_full { flush }`

I think speed is acceptable here if you don't like that.

schneems commented:
The API seems good, i made some thread safety related comments :heart:

ryansmith3136 committed:
simplify comment regarding flush semantics

ryansmith3136 commented:
Great! I think it is time to :shipit:

eric commented:
I just noticed the `@hash.delete(tok)` in `process_hash()` is going to cause you some problems (you don't need it and it will break things).

ryansmith3136 commented:
@eric Great catch. Fixed and previous release was yanked.

https://github.com/ryandotsmith/lpxc/releases/tag/v0.0.7
